### PR TITLE
Bc dspdc 466 upsert

### PR DESCRIPTION
**Pull request by benjamincarlin**
WIP 
And needs tests

### CODE REVIEW

File: src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java [hunk 1]
```diff
@@ -400,6 +397,137 @@ public void addRowIdsToStagingTable(Dataset dataset, String stagingTableName) {
         bigQueryProject.query(sql.toString());
     }
 
+    public List<String> getOverLappingRows(Dataset dataset,
```
Comment (danxmoran) [inline src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java]: Ben and I talked in-person, but for other people looking at this: the APIs will be reworked so they don't try to pull all the row IDs into memory as a list.

**Review by raaidbroad — APPROVED** [review]

File: src/main/resources/application-integrationtest.properties [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # it for integration test
 
-it.jadeApiUrl=https://jade.datarepo-integration.broadinstitute.org
+it.jadeApiUrl=http://localhost:8080
```
Comment (thathert) [inline src/main/resources/application-integrationtest.properties]: You'll want to undo this change.

File: src/main/java/bio/terra/flight/dataset/ingest/DatasetIngestFlight.java [hunk 3]
```diff
@@ -18,13 +21,26 @@ public DatasetIngestFlight(FlightMap inputParameters, Object applicationContext,
...
-        addStep(new IngestInsertIntoDatasetTableStep(datasetService, bigQueryPdao));
+        if (ingestStrategy == IngestRequestModel.StrategyEnum.UPSERT) {
+            addStep(new IngestEvaluateOverlapStep(datasetService, bigQueryPdao));
+            addStep(new IngestSoftDeleteChangedRowsService(datasetService, bigQueryPdao));
+            addStep(new IngestUpsertIntoDatasetTableStep(datasetService, bigQueryPdao));
+        } else if (ingestStrategy == IngestRequestModel.StrategyEnum.APPEND) {
+            addStep(new IngestInsertIntoDatasetTableStep(datasetService, bigQueryPdao));
+        } else {
+            throw new BadRequestException(
+                "The dataset ingest flight expects the ingest request to have a stratgey of  `upsert` or `append`" +
```
Comment (thathert) [inline src/main/java/bio/terra/flight/dataset/ingest/DatasetIngestFlight.java]: typo on `strategy`

File: src/main/java/bio/terra/flight/dataset/ingest/IngestSetupStep.java [hunk 4]
```diff
@@ -55,6 +64,34 @@ public StepResult doStep(FlightContext context) {
...
+                .get()
+                .getPrimaryKey();
+            if (primaryKey.size() < 1) {
+                // TODO: add test
+                throw new BadRequestException(
+                    "The dataset ingest flight expects ingestStrategy `upsert` or `append` but was "
+                        + ingestStrategy.toString());
+            }
+
+            Schema OverlappingTableSchema = bigQueryPdao.buildOverlappingTableSchema();
```
Comment (thathert) [inline src/main/java/bio/terra/flight/dataset/ingest/IngestSetupStep.java]: lowercase variable names: `overlappingTableSchema`

File: src/main/java/bio/terra/flight/dataset/ingest/IngestSetupStep.java [hunk 5]
```diff
@@ -55,6 +64,34 @@ public StepResult doStep(FlightContext context) {
         String sgName = FlightUtils.randomizeNameInfix(baseName, "_st_");
         IngestUtils.putStagingTableName(context, sgName);
 
+        IngestRequestModel ingestRequestModel = IngestUtils.getIngestRequestModel(context);
+        IngestRequestModel.StrategyEnum ingestStrategy = ingestRequestModel.getStrategy();
+
+        if (ingestStrategy == IngestRequestModel.StrategyEnum.UPSERT) {
+            List<Column> primaryKey = dataset
+                .getTableByName(targetTable.getName())
+                .get()
```
Comment (thathert) [inline src/main/java/bio/terra/flight/dataset/ingest/IngestSetupStep.java]: If this is an `Optional<List<Column>>` then we should not always assume that we can `.get()`. If we can't get the target table by name, let's throw a `CorruptMetadataException`.

File: src/main/java/bio/terra/flight/dataset/ingest/IngestUpsertIntoDatasetTableStep.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+        bigQueryPdao.upsertIntoDatasetTable(dataset, targetTable, stagingTableName, overlappingTableName);
+
+        return StepResult.getStepResultSuccess();
+    }
+
+    @Override
+    public StepResult undoStep(FlightContext context) {
+        // We do not need to undo the data insert. BigQuery guarantees that this statement is atomic, so either
+        // of the data will be in the table or we will fail and none of the data is in the table. The
```
Comment (thathert) [inline src/main/java/bio/terra/flight/dataset/ingest/IngestUpsertIntoDatasetTableStep.java:52]: Did some of this comment get cut off?

File: src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java [hunk 7]
```diff
@@ -47,16 +47,10 @@
 import org.springframework.stereotype.Component;
 
 import java.time.Instant;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Comment (thathert) [inline src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java]: Can you expand this? We avoid `.*` imports.

File: src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java [hunk 8]
```diff
@@ -47,16 +47,10 @@
...
 import java.util.stream.Collectors;
 
-import static bio.terra.pdao.PdaoConstant.PDAO_PREFIX;
-import static bio.terra.pdao.PdaoConstant.PDAO_ROW_ID_COLUMN;
-import static bio.terra.pdao.PdaoConstant.PDAO_ROW_ID_TABLE;
-import static bio.terra.pdao.PdaoConstant.PDAO_TABLE_ID_COLUMN;
+import static bio.terra.pdao.PdaoConstant.*;
```
Comment (thathert) [inline src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java]: Same here

File: src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java [hunk 9]
```diff
@@ -400,14 +394,194 @@ public void addRowIdsToStagingTable(Dataset dataset, String stagingTableName) {
...
+         *      OR (TO_JSON_STRING(T.column name) != TO_JSON_STRING(S.column name))
+         */
+        List<String> columnNames = targetTable.getColumns()
+            .stream()
+            .map(Column::getName)
+            .collect(Collectors.toList());
+
+        List<String> naturalKeyColumnNames = dataset
+            .getTableByName(targetTable.getName())
+            .get()
```
Comment (thathert) [inline src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java]: Protect against table not being present like above.

**Review by thathert — APPROVED** [review]
A few style and consistency things below to address, but otherwise this looks ready to go. I'm approving with the assumption that my points will be addressed before merging.

**Pull request merged**